libs/s25main/notifications/NotificationManager.h

```diff
@@ -29,36 +29,34 @@ class NotificationManager
     /// Type of the callback for a Notification(Note)
     template<class T_Note>
     struct NoteCallback;
-    /// Class used as a deleter for shared_ptr to unregister a callback
-    template<class T_Note>
-    class CallbackUnregistrar;
 
 public:
-    NotificationManager();
     ~NotificationManager();
 
     /// Subscribe to a specific notification.
     /// Unsubscribes when the subscription has no references left
     template<class T_Note>
-    Subscription subscribe(std::function<void(T_Note)> callback);
+    Subscription subscribe(std::function<void(const T_Note&)> callback) noexcept;
```

Comment:
should we probably use some existing library for observer pattern instead of implementing it ourselfs? (reactivex, or similar?)

Reply:
I was thinking about this too. But that would add another dependency and likely be slow. We could also use Boost.Signals2 somehow.
This solution has a few advantages:
- It is very fast and needs to (shown by changing list to deque having a measurable impact)
- Pretty small with small compiletime overhead (Boost.Signals2 is header only and as you can imagine large because it needs to be very general)
- Allows passing of arbitrary data and that is the type on which to subscribe which makes it very easy to use and type safe

If we disallow subscribe, unsubscribe and publish during publishing (again) the class would even be simpler.

And TBH: I don't feel like rethinking the usages, subscriptions/callback handling etc. and learning a new framework. This works, has worked and is now even more flexible, so why change it? (CI failures were some spurious, unrelated issues and minor oversights introduced during optimization [e.g deque invalidates iterators but not references, which I missed])

Reply:
what about https://github.com/ddinu/observable ?

Reply:
Looks better (has the possibility to send data) but still: We would still need the "registry" with the mapping of types to signals and the accompanying type erasure (`void*`) or another approach for the signals. Major problem though: Use of std::thread-stuff (mutex and so on) which doesn't work on MinGW
Also as mentioned: The current implementation is tuned to our use case and hence faster than the alternatives.